README.md
# Astronomy
Astronomy is a dichromatic color theme meant for keeping things straightforward
and functional.

Source code starts at neutral, then shifts either warmer to represent value, or
cooler to represent logic.

Equations stay neutral with minimal variance, so separators like strings and
brackets are the only things that stick out.

There are no bright, vibrant colors; you and your extensions are free to use
those for yourself.

![Demonstration screenshot](images/screen1.png)

Also included is a red-green colorblind accessible version of this theme.

![Colorblind demonstration screenshot](images/colorblind.png)

A light version is also included.

![Light theme demonstration screenshot](images/light.png)

## Tips
* Although *any* font is fine, the font featured in the screenshots is [Iosevka Slab](https://github.com/be5invis/Iosevka).
* This theme works well with [Bracket Pair Colorizer 2](https://github.com/CoenraadS/Bracket-Pair-Colorizer-2).

---------

## Building
Use the default gulp task to build the `xxx-color-theme.json` files which will be used by VS Code. Also available is a `watch` task which will automatically build the theme files whenever the template file or palette files are updated.

To add a new theme to be built, modify the `palettes` variable in the `build` function in the gulpfile. Create a new palette file in the `src` directory, e.g. `xyz.json`. Then, add `xyz` as an entry in the `palettes` variable mentioned above. The output theme file will be `astronomy-xyz-color-theme.json` in the `themes` directory. Don't forget to update `package.json` to include the new theme, or else it won't appear as an option in VS Code.

## Palette files
Inside the `src` directory are several small JSON files containing some color definitions, which are then plugged into `_template.json` to create the output theme, which VS Code uses.

* `name`: This populates the "name" field in the output theme file.
* `type`: This populates the "type" field in the output theme file, and whether the type is "light" or not will alter how interpolated colors are generated (see next section)
* `b`: This is the **background** color, used all over the UI.
* `n`: This is the **neutral** color, used as the text color when there's no tinting, but also the foreground color for various UI elements.
* `v1`, `v2`, etc: These are the **value** tints. The higher the number, the stronger the tint.
* `l1`, `l2`, etc: These are the **logic** tints. The higher the number, the stronger the tint.
* `base`: This is the color all of the **interpolated colors** blend towards (see next section).
  * If not specified, this defaults to `#000000` for dark themes, and `#FFFFFF` for light themes.

### Interpolated Colors
Several extra colors are derived from the palette file during the build process, these are **interpolated colors**.

* `v1-75`: This means color `v1` at `75%`. If the theme is dark, lower percentages darken the color. If the theme is light, lower percentages lighten the color. Currently, valid percentages are 25, 33, 50, 67, and 75.
* `negative`: This color is either black if the theme is dark, or `b-75` if the theme is light. Its intended use is for negative space and dividers in the UI.
* You can automatically generate `v1`, `v2`, `l1`, `l2`, etc, by instead specifying a `v` color and an `l` color. The various tints will be interpolated by blending with `n`.

You are free to override any interpolated colors by specifying them in the palette file directly. For example, defining `v1-25` in the palette file will override the auto-generated `v1-25`.

CHANGELOG.md
# Change Log

# 0.5.4

### Added
- For Markdown:
  - Highlighting and decoration for \~\~strikethrough text\~\~.
  - Highlighting for text between \`backticks\`.

# 0.5.3

### Added
- TS/JS `in`, `of`, `delete`, `typeof` highlighting.

# 0.5.2

### Added
- Additional optional blending options
  - Optionally specify `v` and `l` tints to have the `x1`, `x2`, `x3` tints
    automatically generated from blending with `n`.
  - Optionally specify the `base` color all the interpolated colors blend with.
- Coloring for quick picker (command bar) and tooltips.
- Coloring for links and inline code in Markdown previewer.

# 0.5.1

### Modified
- Increased contrast of tab bar.

# 0.5.0

### Added
- Very basic support for semantic highlighting.
  - Largely redundant because fallback TextMate scopes, but still added some
    ones unique to semantic highlighting.
  - Tested in C++ only so far. 
- Top border highlight for tabs.
- MIT Licensing for source code.

# 0.4.0

### Added
- Highlighting for RegEx strings in TS/JS and Python.

# 0.3.2

### Modified
- "Remote Host" button on status bar adjusted to still be distinct but not stand
out as much.

# 0.3.1

### Modified
- "Negative" color for light themes is now b-75 instead of b, which allows the
activity bar to have better contrast.
- Active tab on activity bar now has the panel BG color, to make it look like
a tab.
- Adjusted inputOption colors to make selected options more visible. (Check the
buttons in the "find" dialog's input box)


# 0.3.0

### Added
- Light theme

### Modified
- UI background/foreground colors now come from palette files
- Palette files specify output theme's "name" property and "type" property
- Color interpolator only processes values starting with '#' (won't try to
interpolate "name" or "type", which aren't colors)

### Notes
- Background and foreground colors around the UI have changed very slightly as a
side effect of switching to interpolated values instead of the hardcoded
constants from before

# 0.2.0

### Added
- Red-Green colorblind accessible theme

### Modified
- Themes are now built from template


# 0.1.0

### Added
- Initial theming for UI
- Initial highlighting for source code
  - Tuned for TypeScript, SCSS, and HTML

### Notes
- Compatibility (unthoroughly) checked for:
  - C++
  - PHP
  - Python
  - Markdown
- Git and diff theming is not present

gulpfile.js
var gulp = require('gulp');
var fs = require('fs');
var mustache = require('mustache');

function getColorFromHex(colorHex) {
    // Convert from #RRGGBB... to decimal values [R, G, B, ...]
    channels = [];
    for (var i = 0; i < 4; i++) {
        if ((i * 2) >= colorHex.length - 1) { break; }
        channels.push(parseInt(colorHex.substr(1 + (i * 2),2), 16));
    }
    return channels;
}

function getHexFromColor(color) {
    // Convert from [R, G, B, ...] to hex #RRGGBB...
    var out = "#";
    color.forEach(currChn => {
        var currHex = currChn.toString(16);
        out += (currHex.length < 2 ? "0" : "") + currHex;
    });
    return out;
}

function scaleColorHex(color, target, percent) {
    // Blend color to target
    var channels = getColorFromHex(color);
    for (var i = 0; i < 3; i++) {
        channels[i] = (((channels[i] - target[i]) * percent | 0) + target[i]);
    }
    return getHexFromColor(channels);
}

function build(done) {
    // Define color variants here
    var palettes = ["standard", "colorblind", "miami", "light-standard"];
    var templateContent = fs.readFileSync("src/_template.json").toString();
    var shades = [75, 67, 50, 33, 25];
    var base;
    palettes.forEach(currPalette => {
        // Parse current palette
        var paletteData = JSON.parse(fs.readFileSync(`src/${currPalette}.json`));
        // Get the base color if it exists, or use default if it doesn't.
        if (paletteData["base"] !== undefined) {
            base = getColorFromHex(paletteData["base"]);
        } else {
            // Use default base depending on whether theme is light or dark.
            if (paletteData["type"] == "light") {
                base = getColorFromHex("#FFFFFF");
            } else {
                base = getColorFromHex("#000000");
            }
        }
        // If "l" and/or "v" exist, interpolate between them to make the different tints.
        ["l", "v"].forEach(tint => {
            var neutral = getColorFromHex(paletteData["n"]);
            if (paletteData[tint] !== undefined) {
                for (var i = 1; i < 4; i++) {
                    var tintName = tint + i.toString();
                    if (paletteData[tintName] === undefined) {
                        paletteData[tintName] = scaleColorHex(paletteData[tint], neutral, i/3);
                    }
                }
            }
        });
        // Generate special v4 tint if we were using "v" and "l" to generate the tints.
        if ((paletteData["v4"] === undefined) && (paletteData["v"] !== undefined) && (paletteData["l"] !== undefined)) {
            paletteData["v4"] = scaleColorHex(paletteData["v"], getColorFromHex(paletteData["l"]), 3/4);
        }
        // Create the different shades of each color. Detect which keys are
        // colors by checking if the first character of the value is '#'.
        var toShade = Object.keys(paletteData).filter(key => paletteData[key].charAt(0) == "#");
        toShade.forEach(hue => {
            shades.forEach(currShade => {
                var keyName = hue + "-" + currShade;
                // Don't overwrite existing values (allows you to specify something
                // explicitly in the palette file)
                if (paletteData[keyName] === undefined) {
                    paletteData[keyName] = scaleColorHex(paletteData[hue], base, currShade / 100);
                }
            })
        })
        // Create "negative space" color
        if (paletteData["negative"] === undefined) {
            if (paletteData["type"] == "light") {
                paletteData["negative"] = paletteData["b-75"];
            } else {
                paletteData["negative"] = getHexFromColor(base);
            }
        }
        // Render template with palette
        var outContent = mustache.render(templateContent, paletteData);
        // Create the result file
        fs.writeFileSync(`themes/astronomy-${currPalette}-color-theme.json`, outContent);
    });
    done();
};

function watch() {
    gulp.series(build)();
    gulp.watch('src/**/*.json', build);
};

exports.default = build;
exports.watch = watch;